src/main.c

```diff
@@ -1,69 +1,123 @@
+         */
+        current->pnext = next;
     } else {
-        //This doesn't behave how I expected it to
-        #warning attempting to add 'next' to a node that already has next
+        /**
+         * Preprocessor warnings occur whenever the preprocessor sees them. This
+         * occurs before compilation, and has nothing to do with actual source
+         * code. This is why preprocessor declarations should always be put at
+         * the beginning of a line.
+         * #warning creates a compilation warning—not a runttime one
+         */
+        fprintf(stderr, "error: "
+                "attempting to add 'next' to a node that already has next\n");
     }
 }
-*/
 
-int main() {
-    //TODO get data from user
-    int data[] = {1, 2, 3, 4, 5, 6, 7, 8, 9}, i = 0;
-    //will this work
-    struct node *head = malloc(sizeof(struct node)), *current = head;
-    //or do we need this
-    //struct node *head = (struct node *) malloc(sizeof(struct node));
+/**
+ * Call function main as follows: ./main 1 2 3 4 5 6 7 8...
+ */
+int main(int argc, char **argv) {
+    if (argc == 1) {
+        fprintf(stderr, "Need at least 1 argument!", argc);
+        exit(EXIT_FAILURE);
+    }
+
+    fprintf(stderr, "main called with %d arguments\n", argc);
+    /**
+     * Create an array to hold the arguments (not really necessary, but for
+     * demonstration how to use calloc.
+     * calloc is better than malloc since it automatically initializes its data
+     * to `0`. with malloc the data is uninitialized!
+     */
+    int length = argc-1;
+    int *data = calloc(sizeof(int), argc);
+    /* Get user input. note: first argument is at argv[1] */
+    for (int i = 0; i < length; ++i) {
+        data[i] = atoi(argv[i+1]);
+    }
 
-    int data_length = sizeof(data) / sizeof(data[0]);
+    /**
+     * The original way this was done was bad for two reasons.
+     * First of all, if the list were empty, then there would only be one node,
+     * and it would be completely uninitialized.
+     *
+     * Furthermore, for any other size, there will be one extra element at the
+     * end of the list (since the loop creates a node every loop, and there's
+     * one at the beginning, there will always be n+1 nodes).
+     * In either case, the last node is uninitialized, which means that there's
+     * no `NULL` at the end of the list which is really bad.
+     *
+     * Instead, we will set up the list differently.
+     */
+    struct node *head, *current;
+    head = current = NULL;
 
-    //set up our list
-    while(i < data_length) {
-        current -> item = data[i]; //set data for current node
-        //make a new node and point to it
-        current -> pnext = malloc(sizeof(struct node));
-        //point current to that new node
-        current = current -> pnext;
-        i++;
+    /* set up our list */
+    for (int i = 0; i < length; ++i) {
+        /* allocate the current node */
+        if (!current) {
```

Comment:
how does using `!` on a pointer work? It's a new concept to me using boolean ops on non boolean values. Does it check for `NULL`?

Reply:
C doesn't have booleans.
`if` and friends actually just check for `0`. `true` is just a macro which resolves to `1` and `false` to `0`.
therefore
`if (!current)` is equivalent to `if (current == 0)` which is the same as `if (current == NULL)`

it's just a nice thing you can do to check if pointers are valid, so instead of
```
if (p != NULL) {
return *p;
}
```
you can just do
```
if (p) {
return *p;
}
```